src/main/src/main/java/org/geoserver/security/impl/LayerGroupContainmentCache.java

```diff
@@ -0,0 +1,426 @@
+        List<LayerGroupInfo> groups = catalog.getLayerGroups();
+
+        // first populate the basic structure
+        for (LayerGroupInfo lg : groups) {
+            addGroupInfo(lg);
+        }
+
+        // now populate the containment structure
+        for (LayerGroupInfo lg : groups) {
+            registerContainedGroups(lg);
+        }
+
+    }
+
+    private void registerContainedGroups(LayerGroupInfo lg) {
+        lg.getLayers().stream().filter(IS_GROUP).forEach(p -> {
+            String containerId = lg.getId();
+            String containedId = p.getId();
+            LayerGroupSummary container = groupCache.get(containerId);
+            LayerGroupSummary contained = groupCache.get(containedId);
+            if (container != null && contained != null) {
+                contained.containerGroups.add(container);
+            }
+        });
+    }
+
+    private void addGroupInfo(LayerGroupInfo lg) {
+        LayerGroupSummary groupData = new LayerGroupSummary(lg);
+        groupCache.put(lg.getId(), groupData);
+        lg.getLayers().stream().filter(IS_LAYER).forEach(p -> {
+            String id = ((LayerInfo) p).getResource().getId();
+            Set<LayerGroupSummary> containers = resourceContainmentCache.computeIfAbsent(id,
+                    CONCURRENT_SET_BUILDER);
+            containers.add(groupData);
+        });
+    }
+
+    private void clearGroupInfo(LayerGroupInfo lg) {
+        LayerGroupSummary data = groupCache.remove(lg.getId());
+        // clear the resource containment cache
+        lg.getLayers().stream().filter(IS_LAYER).forEach(p -> {
+            String rid = ((LayerInfo) p).getResource().getId();
+            synchronized (rid) {
+                Set<LayerGroupSummary> containers = resourceContainmentCache.get(rid);
+                if (containers != null) {
+                    containers.remove(data);
+                }
+            }
+        });
+        // this group does not contain anything anymore, remove from containment
+        for (LayerGroupSummary d : groupCache.values()) {
+            d.containerGroups.remove(lg.getId());
+        }
+    }
+
+    /**
+     * Returns all groups containing directly or indirectly containing the resource
+     * 
+     * @param resource
+     * @return
+     */
+    public Collection<LayerGroupSummary> getContainerGroupsFor(ResourceInfo resource) {
+        String id = resource.getId();
+        Set<LayerGroupSummary> groups = resourceContainmentCache.get(id);
+        if (groups == null) {
+            return Collections.emptyList();
+        }
+        Set<LayerGroupSummary> result = new HashSet<>();
+        for (LayerGroupSummary lg : groups) {
+            collectContainers(lg, result);
+        }
+        return result;
+    }
+
+    /**
+     * Returns all groups containing directly or indirectly the specified group, and relevant for security (e.g., anything but
+     * {@link LayerGroupInfo.Mode#SINGLE} ones
+     * 
```

Comment:
So opaque container groups are relevant for security because they can actually hide a layer right ?
By the way the method above should probably also say that the single groups will not be considered since they are filtered in the ``collectContainers`` method :)